### PR TITLE
feat: add support for custom gas token

### PR DESCRIPTION
**Pull request by spsjvc**
### L2Network

* Added a new property `nativeToken`

### AssetBridger

* Added a new property `nativeToken`
* Added a new property `nativeTokenIsEth` which is a boolean that represents whether the Arbitrum chain that is being used with the bridger uses ETH as the native token.

### EthBridger

Added `getApproveGasTokenRequest` and `approveGasToken` to `EthBridger`, which help approve the custom gas token to be spent by the Inbox. When adding support for a custom gas token chain, the user only needs to call this before doing the regular deposit, while the rest of the flow stays the same. The bridger object will pick between 2 separate code paths depending on the value of `nativeTokenIsEth`.

These methods will throw if used on a chain that uses ETH as the gas token.

### Erc20Bridger

Added `getApproveGasTokenRequest` and `approveGasToken` to `EthBridger`, which help approve the custom gas token to be spent by the appropriate Gateway. When adding support for a custom gas token chain, the user only needs to call this before doing the regular deposit, while the rest of the flow stays the same. The bridger object will pick between 2 separate code paths depending on the value of `nativeTokenIsEth`.

These methods will throw if used on a chain that uses ETH as the gas token.

### L1ToL2MessageCreator

Updated `getTicketCreationRequest` to support custom gas token chains and have different calldata in that case.

### CODE REVIEW

**Review by brtkx — APPROVED** [review]

File: src/lib/assetBridger/erc20Bridger.ts [hunk 1]
```diff
@@ -507,6 +549,64 @@ export class Erc20Bridger extends AssetBridger<
...
+   */
+  private getDepositRequestOutboundTransferInnerData(
+    depositParams: OmitTyped<L1ToL2MessageGasParams, 'deposit'>
+  ) {
+    if (!this.nativeTokenIsEth) {
+      return defaultAbiCoder.encode(
+        ['uint256', 'bytes', 'uint256'],
+        [
+          // maxSubmissionCost
+          constants.Zero,
```
[the remark this answers is not in the document]
Comment (spsjvc) [reply src/lib/assetBridger/erc20Bridger.ts:588]: Good call 👍

File: scripts/genNetwork.ts [hunk 2]
```diff
@@ -1,25 +1,85 @@
...
+) {
+  // we need to add partnerChainIDs to the L2 network
+  l2Network.partnerChainIDs = l3Network ? [l3Network.chainID] : []
+  l2Network.blockTime = ARB_MINIMUM_BLOCK_TIME_IN_SECONDS
+
+  // native token for l3
+  if (l3Network && l2Provider) {
+    l3Network.partnerChainIDs = []
+    l3Network.blockTime = ARB_MINIMUM_BLOCK_TIME_IN_SECONDS
+    try {
```
[the remark this answers is not in the document]
Comment (spsjvc) [reply scripts/genNetwork.ts:51]: This should still work, as the SDK assumes chain is ETH-based if the `nativeToken` property is either `undefined` or the zero address.

File: scripts/fundCustomFeeToken.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    console.error('No local networks found')
+    //TODO: get token address from deployment somehow
+  }
+  const nativeTokenAddress = localNetworks?.l2Network.nativeToken
+  if (!nativeTokenAddress || nativeTokenAddress === undefined) {
+    console.error('No native token found')
+    return
+  }
+
+  await fundL1CustomFeeToken(nativeTokenAddress)
```
[the remark this answers is not in the document]
Comment (spsjvc) [reply scripts/fundCustomFeeToken.ts:16]: Don't think this script is used anymore, so I'll remove

File: package.json [hunk 4]
```diff
@@ -54,8 +54,8 @@
     "ethers": "^5.1.0"
   },
   "devDependencies": {
-    "@arbitrum/nitro-contracts": "1.1.0",
-    "@arbitrum/token-bridge-contracts": "1.2.0",
+    "@arbitrum/nitro-contracts": "^1.1.0",
```
Comment (gzeoneth) [inline package.json:57]: let's bump to ^1.1.1 to get rid of sol2uml
https://github.com/OffchainLabs/nitro-contracts/releases/tag/v1.1.1

File: package.json [hunk 5]
```diff
@@ -54,8 +54,8 @@
...
+    "@arbitrum/token-bridge-contracts": "^1.2.0",
```
Comment (gzeoneth) [inline package.json:58]: let's bump to ^1.2.1 to get rid of sol2uml
https://github.com/OffchainLabs/token-bridge-contracts/releases/tag/v1.2.1

File: src/lib/assetBridger/erc20Bridger.ts [hunk 6]
```diff
@@ -696,6 +788,11 @@ export class Erc20Bridger extends AssetBridger<
       // in the future we want to do proper estimation here
       /* eslint-disable @typescript-eslint/no-unused-vars */
       estimateL1GasLimit: async (l1Provider: Provider) => {
+        if (!this.nativeTokenIsEth) {
+          // measured 172867 - add some padding
+          return BigNumber.from(200000)
```
Comment (gzeoneth) [inline src/lib/assetBridger/erc20Bridger.ts:793]: this probably won't work for L3->L2 withdrawal

File: src/lib/assetBridger/erc20Bridger.ts [hunk 7]
```diff
@@ -706,7 +803,7 @@ export class Erc20Bridger extends AssetBridger<
         const isWeth = await this.isWethGateway(l1GatewayAddress, l1Provider)
 
         // measured 157421 - add some padding
-        return isWeth ? BigNumber.from(180000) : BigNumber.from(160000)
+        return isWeth ? BigNumber.from(190000) : BigNumber.from(160000)
```
Comment (gzeoneth) [inline src/lib/assetBridger/erc20Bridger.ts:806]: this probably won't work for L3->L2 withdrawal

File: package.json (hunk 4 above)
Comment (spsjvc) [reply package.json]: bumped in https://github.com/OffchainLabs/arbitrum-sdk/pull/310/commits/af5586fc54fefc204d86f0eceef671f42ebd8fcf

File: src/lib/assetBridger/erc20Bridger.ts (hunk 6 above)
Comment (spsjvc) [reply src/lib/assetBridger/erc20Bridger.ts:793]: bumped in https://github.com/OffchainLabs/arbitrum-sdk/pull/310/commits/af5586fc54fefc204d86f0eceef671f42ebd8fcf

File: package.json (hunk 5 above)
Comment (spsjvc) [reply package.json]: bumped in https://github.com/OffchainLabs/arbitrum-sdk/pull/310/commits/af5586fc54fefc204d86f0eceef671f42ebd8fcf

**Review by gzeoneth — APPROVED** [review]
LGTM
recommend to add a comment regarding the hardcoded gas: the correct way to do is use NodeInterace.gasEstimateL1Component to calculate the L1 component + hardcoded L2 component

**Pull request merged**